exercises/1-User_Mode/challenge_101/solution/101_example.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Filename: 101_source.c
+ * Brief: All exercise code for User Mode challenge 101 goes here.
+ * Details:
+ *     - Define the contents of initSigHandlers()
+ *     - Catch all possible signals
+ *     - Print "Ignoring signal: X" to stderr when a signal is caught
+ *     - Feel free to define your "signal handling function" in this source file
+ */
+
+#include <errno.h>
+#include <signal.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#define NAP_TIME (int)5  // Number of seconds to sleep
+
+
+/*
+ * Function: initSigHandlers
+ * Description: Install "logging" signal handlers for all possible signals
+ */
+void initSigHandlers(void);
+
+
+int main(void)
+{
+    // Prepare stdout
+    setvbuf(stdout, NULL, _IONBF, 0);  // Make stdout unbuffered
+    setvbuf(stderr, NULL, _IONBF, 0);  // Make stderr unbuffered
+
+    // Initialize Signal Handlers
+    initSigHandlers();
+
+    // Wait
+    printf("Waiting for a signal...");
+    while(1)
+    {
+        sleep(NAP_TIME);
+        printf(".");
+    }
+    printf("Done\n");
+    return 0;
+}
+
+
+/*
+ * Function: ignore
+ * Description: Prints the ignored signal number to stderr
+ *
+ * signalNo: Incoming signal value
+ */
+void ignore(int signalNo, siginfo_t *info, void *context)
+{
+    fprintf(stderr, "Ignoring signal: %d\n", signalNo);
```

Comment:
From the Linux Programming Interface:
> Not all system calls and library functions can be safely called from a signal handler.
> Although `printf()` is not async-signal-safe, [it is frequently used in example signal handlers]]

Safe functions must be "reentrant" and "async-signal-safe".
`printf()` is not safe for use in a signal handler.  Refactor this to use `write()` instead.

Reply:
Refactored.